Mathlib/RingTheory/IntegralClosure/Algebra/Ideal.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+module
+
+public import Mathlib.RingTheory.IntegralClosure.Algebra.Basic
+
+/-!
+
+# Integrality over ideals
+
+## Main results
+- `Polynomial.exists_monic_aeval_eq_zero_forall_mem_of_mem_map`:
+  If `S` is an integral `R`-algebra, and `I` is an ideal of `R`,
+  then any `x ∈ IS` is integral over `I`, i.e. it is a root
+  of some monic polynomial in `R[X]` whose non-leading coefficients are in `I`.
+
+-/
+
+@[expose] public section
+
+namespace Polynomial
+
+variable {R S : Type*} [CommRing R] [CommRing S] [Algebra R S]
+
+lemma coeff_mem_pow_of_mem_adjoin_C_mul_X {R : Type*} [CommRing R]
+    {I : Ideal R} {P : R[X]} (hP : P ∈ Algebra.adjoin R { C r * X | r ∈ I }) (i : ℕ) :
+    P.coeff i ∈ I ^ i := by
+  induction hP using Algebra.adjoin_induction generalizing i with
+  | mem x hx =>
+    obtain ⟨r, hrI, rfl⟩ := hx
+    simp +contextual [coeff_X, apply_ite, hrI, @eq_comm _ 1]
+  | algebraMap r => simp +contextual [coeff_C, apply_ite]
+  | add x y hx hy _ _ => aesop
+  | mul x y _ _ hx hy =>
+    rw [coeff_mul]
+    refine sum_mem fun ⟨j₁, j₂⟩ hj ↦ ?_
+    obtain rfl : j₁ + j₂ = i := by simpa using hj
+    exact pow_add I j₁ j₂ ▸ Ideal.mul_mem_mul (hx _) (hy _)
+
+attribute [local instance] Polynomial.algebra in
+lemma exists_monic_aeval_eq_zero_forall_mem_pow_of_isIntegral
+    {I : Ideal R} {x : S}
+    (hx : IsIntegral (Algebra.adjoin R { C r * X | r ∈ I }) (C x * X)) :
+    ∃ p : R[X], p.Monic ∧ aeval x p = 0 ∧ ∀ i, p.coeff i ∈ I ^ (p.natDegree - i) := by
```

Comment:
This seems to be a stronger condition on the coefficients then the one mentioned above.